// json-editor-vue/dist/json-editor-vue.mjs
import { Mode as nn, createJSONEditor as en } from "vanilla-jsoneditor";
import { isVue3 as St, defineComponent as on, getCurrentInstance as an, ref as pt, watchEffect as un, computed as Vt, onUnmounted as cn, onMounted as fn, watch as dt, unref as sn, h as ln } from "vue-demi";
const pn = /"(?:_|\\u0{2}5[Ff]){2}(?:p|\\u0{2}70)(?:r|\\u0{2}72)(?:o|\\u0{2}6[Ff])(?:t|\\u0{2}74)(?:o|\\u0{2}6[Ff])(?:_|\\u0{2}5[Ff]){2}"\s*:/, dn = /"(?:c|\\u0063)(?:o|\\u006[Ff])(?:n|\\u006[Ee])(?:s|\\u0073)(?:t|\\u0074)(?:r|\\u0072)(?:u|\\u0075)(?:c|\\u0063)(?:t|\\u0074)(?:o|\\u006[Ff])(?:r|\\u0072)"\s*:/, vn = /^\s*["[{]|^\s*-?\d{1,16}(\.\d{1,17})?([Ee][+-]?\d+)?\s*$/;
function bn(t, r) {
  if (t === "__proto__" || t === "constructor" && r && typeof r == "object" && "prototype" in r) {
    yn(t);
    return;
  }
  return r;
}
function yn(t) {
  console.warn(`[destr] Dropping "${t}" key to prevent prototype pollution.`);
}
function Q(t, r = {}) {
  if (typeof t != "string")
    return t;
  const n = t.trim();
  if (
    // eslint-disable-next-line unicorn/prefer-at
    t[0] === '"' && t.endsWith('"') && !t.includes("\\")
  )
    return n.slice(1, -1);
  if (n.length <= 9) {
    const e = n.toLowerCase();
    if (e === "true")
      return !0;
    if (e === "false")
      return !1;
    if (e === "undefined")
      return;
    if (e === "null")
      return null;
    if (e === "nan")
      return Number.NaN;
    if (e === "infinity")
      return Number.POSITIVE_INFINITY;
    if (e === "-infinity")
      return Number.NEGATIVE_INFINITY;
  }
  if (!vn.test(t)) {
    if (r.strict)
      throw new SyntaxError("[destr] Invalid JSON");
    return t;
  }
  try {
    if (pn.test(t) || dn.test(t)) {
      if (r.strict)
        throw new Error("[destr] Possible prototype pollution");
      return JSON.parse(t, bn);
    }
    return JSON.parse(t);
  } catch (e) {
    if (r.strict)
      throw e;
    return t;
  }
}
function hn(t, r = {}) {
  return Q(t, { ...r, strict: !0 });
}
var gn = typeof global == "object" && global && global.Object === Object && global, jn = typeof self == "object" && self && self.Object === Object && self, Sr = gn || jn || Function("return this")(), nt = Sr.Symbol, Tr = Object.prototype, mn = Tr.hasOwnProperty, _n = Tr.toString, R = nt ? nt.toStringTag : void 0;
function On(t) {
  var r = mn.call(t, R), n = t[R];
  try {
    t[R] = void 0;
    var e = !0;
  } catch {
  }
  var i = _n.call(t);
  return e && (r ? t[R] = n : delete t[R]), i;
}
var wn = Object.prototype, An = wn.toString;
function xn(t) {
  return An.call(t);
}
var Sn = "[object Null]", Tn = "[object Undefined]", Rt = nt ? nt.toStringTag : void 0;
function En(t) {
  return t == null ? t === void 0 ? Tn : Sn : Rt && Rt in Object(t) ? On(t) : xn(t);
}
function In(t) {
  return t != null && typeof t == "object";
}
var Fn = "[object Symbol]";
function Pn(t) {
  return typeof t == "symbol" || In(t) && En(t) == Fn;
}
var kn = /\s/;
function zn(t) {
  for (var r = t.length; r-- && kn.test(t.charAt(r)); )
    ;
  return r;
}
var Cn = /^\s+/;
function Nn(t) {
  return t && t.slice(0, zn(t) + 1).replace(Cn, "");
}
function gt(t) {
  var r = typeof t;
  return t != null && (r == "object" || r == "function");
}
var Wt = NaN, $n = /^[-+]0x[0-9a-f]+$/i, Ln = /^0b[01]+$/i, Bn = /^0o[0-7]+$/i, Mn = parseInt;
function Gt(t) {
  if (typeof t == "number")
    return t;
  if (Pn(t))
    return Wt;
  if (gt(t)) {
    var r = typeof t.valueOf == "function" ? t.valueOf() : t;
    t = gt(r) ? r + "" : r;
  }
  if (typeof t != "string")
    return t === 0 ? t : +t;
  t = Nn(t);
  var n = Ln.test(t);
  return n || Bn.test(t) ? Mn(t.slice(2), n ? 2 : 8) : $n.test(t) ? Wt : +t;
}
var vt = function() {
  return Sr.Date.now();
}, Dn = "Expected a function", Un = Math.max, Vn = Math.min;
function Rn(t, r, n) {
  var e, i, o, a, u, c, f = 0, s = !1, d = !1, v = !0;
  if (typeof t != "function")
    throw new TypeError(Dn);
  r = Gt(r) || 0, gt(n) && (s = !!n.leading, d = "maxWait" in n, o = d ? Un(Gt(n.maxWait) || 0, r) : o, v = "trailing" in n ? !!n.trailing : v);
  function h(g) {
    var x = e, V = i;
    return e = i = void 0, f = g, a = t.apply(V, x), a;
  }
  function p(g) {
    return f = g, u = setTimeout(l, r), s ? h(g) : a;
  }
  function j(g) {
    var x = g - c, V = g - f, Ut = r - x;
    return d ? Vn(Ut, o - V) : Ut;
  }
  function A(g) {
    var x = g - c, V = g - f;
    return c === void 0 || x >= r || x < 0 || d && V >= o;
  }
  function l() {
    var g = vt();
    if (A(g))
      return y(g);
    u = setTimeout(l, j(g));
  }
  function y(g) {
    return u = void 0, v && e ? h(g) : (e = i = void 0, a);
  }
  function w() {
    u !== void 0 && clearTimeout(u), f = 0, e = c = i = u = void 0;
  }
  function b() {
    return u === void 0 ? a : y(vt());
  }
  function O() {
    var g = vt(), x = A(g);
    if (e = arguments, i = this, c = g, x) {
      if (u === void 0)
        return p(c);
      if (d)
        return clearTimeout(u), u = setTimeout(l, r), h(c);
    }
    return u === void 0 && (u = setTimeout(l, r)), a;
  }
  return O.cancel = w, O.flush = b, O;
}
var Er = typeof global == "object" && global && global.Object === Object && global, Wn = typeof self == "object" && self && self.Object === Object && self, P = Er || Wn || Function("return this")(), I = P.Symbol, Ir = Object.prototype, Gn = Ir.hasOwnProperty, Jn = Ir.toString, W = I ? I.toStringTag : void 0;
function qn(t) {
  var r = Gn.call(t, W), n = t[W];
  try {
    t[W] = void 0;
    var e = !0;
  } catch {
  }
  var i = Jn.call(t);
  return e && (r ? t[W] = n : delete t[W]), i;
}
var Kn = Object.prototype, Yn = Kn.toString;
function Hn(t) {
  return Yn.call(t);
}
var Xn = "[object Null]", Qn = "[object Undefined]", Jt = I ? I.toStringTag : void 0;
function B(t) {
  return t == null ? t === void 0 ? Qn : Xn : Jt && Jt in Object(t) ? qn(t) : Hn(t);
}
function F(t) {
  return t != null && typeof t == "object";
}
var Zn = "[object Symbol]";
function Tt(t) {
  return typeof t == "symbol" || F(t) && B(t) == Zn;
}
function te(t, r) {
  for (var n = -1, e = t == null ? 0 : t.length, i = Array(e); ++n < e; )
    i[n] = r(t[n], n, t);
  return i;
}
var S = Array.isArray, qt = I ? I.prototype : void 0, Kt = qt ? qt.toString : void 0;
function Fr(t) {
  if (typeof t == "string")
    return t;
  if (S(t))
    return te(t, Fr) + "";
  if (Tt(t))
    return Kt ? Kt.call(t) : "";
  var r = t + "";
  return r == "0" && 1 / t == -1 / 0 ? "-0" : r;
}
function k(t) {
  var r = typeof t;
  return t != null && (r == "object" || r == "function");
}
function Et(t) {
  return t;
}
var re = "[object AsyncFunction]", ne = "[object Function]", ee = "[object GeneratorFunction]", oe = "[object Proxy]";
function It(t) {
  if (!k(t))
    return !1;
  var r = B(t);
  return r == ne || r == ee || r == re || r == oe;
}
var bt = P["__core-js_shared__"], Yt = function() {
  var t = /[^.]+$/.exec(bt && bt.keys && bt.keys.IE_PROTO || "");
  return t ? "Symbol(src)_1." + t : "";
}();
function ie(t) {
  return !!Yt && Yt in t;
}
var ae = Function.prototype, ue = ae.toString;
function M(t) {
  if (t != null) {
    try {
      return ue.call(t);
    } catch {
    }
    try {
      return t + "";
    } catch {
    }
  }
  return "";
}
var ce = /[\\^$.*+?()[\]{}|]/g, fe = /^\[object .+?Constructor\]$/, se = Function.prototype, le = Object.prototype, pe = se.toString, de = le.hasOwnProperty, ve = RegExp(
  "^" + pe.call(de).replace(ce, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"
);
function be(t) {
  if (!k(t) || ie(t))
    return !1;
  var r = It(t) ? ve : fe;
  return r.test(M(t));
}
function ye(t, r) {
  return t == null ? void 0 : t[r];
}
function D(t, r) {
  var n = ye(t, r);
  return be(n) ? n : void 0;
}
var jt = D(P, "WeakMap"), Ht = Object.create, he = /* @__PURE__ */ function() {
  function t() {
  }
  return function(r) {
    if (!k(r))
      return {};
    if (Ht)
      return Ht(r);
    t.prototype = r;
    var n = new t();
    return t.prototype = void 0, n;
  };
}();
function ge(t, r, n) {
  switch (n.length) {
    case 0:
      return t.call(r);
    case 1:
      return t.call(r, n[0]);
    case 2:
      return t.call(r, n[0], n[1]);
    case 3:
      return t.call(r, n[0], n[1], n[2]);
  }
  return t.apply(r, n);
}
function je(t, r) {
  var n = -1, e = t.length;
  for (r || (r = Array(e)); ++n < e; )
    r[n] = t[n];
  return r;
}
var me = 800, _e = 16, Oe = Date.now;
function we(t) {
  var r = 0, n = 0;
  return function() {
    var e = Oe(), i = _e - (e - n);
    if (n = e, i > 0) {
      if (++r >= me)
        return arguments[0];
    } else
      r = 0;
    return t.apply(void 0, arguments);
  };
}
function Ae(t) {
  return function() {
    return t;
  };
}
var et = function() {
  try {
    var t = D(Object, "defineProperty");
    return t({}, "", {}), t;
  } catch {
  }
}(), xe = et ? function(t, r) {
  return et(t, "toString", {
    configurable: !0,
    enumerable: !1,
    value: Ae(r),
    writable: !0
  });
} : Et, Se = we(xe);
function Te(t, r) {
  for (var n = -1, e = t == null ? 0 : t.length; ++n < e && r(t[n], n, t) !== !1; )
    ;
  return t;
}
var Ee = 9007199254740991, Ie = /^(?:0|[1-9]\d*)$/;
function Ft(t, r) {
  var n = typeof t;
  return r = r ?? Ee, !!r && (n == "number" || n != "symbol" && Ie.test(t)) && t > -1 && t % 1 == 0 && t < r;
}
function ut(t, r, n) {
  r == "__proto__" && et ? et(t, r, {
    configurable: !0,
    enumerable: !0,
    value: n,
    writable: !0
  }) : t[r] = n;
}
function H(t, r) {
  return t === r || t !== t && r !== r;
}
var Fe = Object.prototype, Pe = Fe.hasOwnProperty;
function Pr(t, r, n) {
  var e = t[r];
  (!(Pe.call(t, r) && H(e, n)) || n === void 0 && !(r in t)) && ut(t, r, n);
}
function kr(t, r, n, e) {
  var i = !n;
  n || (n = {});
  for (var o = -1, a = r.length; ++o < a; ) {
    var u = r[o], c = e ? e(n[u], t[u], u, n, t) : void 0;
    c === void 0 && (c = t[u]), i ? ut(n, u, c) : Pr(n, u, c);
  }
  return n;
}
var Xt = Math.max;
function ke(t, r, n) {
  return r = Xt(r === void 0 ? t.length - 1 : r, 0), function() {
    for (var e = arguments, i = -1, o = Xt(e.length - r, 0), a = Array(o); ++i < o; )
      a[i] = e[r + i];
    i = -1;
    for (var u = Array(r + 1); ++i < r; )
      u[i] = e[i];
    return u[r] = n(a), ge(t, this, u);
  };
}
function ze(t, r) {
  return Se(ke(t, r, Et), t + "");
}
var Ce = 9007199254740991;
function Pt(t) {
  return typeof t == "number" && t > -1 && t % 1 == 0 && t <= Ce;
}
function ct(t) {
  return t != null && Pt(t.length) && !It(t);
}
function Ne(t, r, n) {
  if (!k(n))
    return !1;
  var e = typeof r;
  return (e == "number" ? ct(n) && Ft(r, n.length) : e == "string" && r in n) ? H(n[r], t) : !1;
}
function zr(t) {
  return ze(function(r, n) {
    var e = -1, i = n.length, o = i > 1 ? n[i - 1] : void 0, a = i > 2 ? n[2] : void 0;
    for (o = t.length > 3 && typeof o == "function" ? (i--, o) : void 0, a && Ne(n[0], n[1], a) && (o = i < 3 ? void 0 : o, i = 1), r = Object(r); ++e < i; ) {
      var u = n[e];
      u && t(r, u, e, o);
    }
    return r;
  });
}
var $e = Object.prototype;
function kt(t) {
  var r = t && t.constructor, n = typeof r == "function" && r.prototype || $e;
  return t === n;
}
function Le(t, r) {
  for (var n = -1, e = Array(t); ++n < t; )
    e[n] = r(n);
  return e;
}
var Be = "[object Arguments]";
function Qt(t) {
  return F(t) && B(t) == Be;
}
var Cr = Object.prototype, Me = Cr.hasOwnProperty, De = Cr.propertyIsEnumerable, ot = Qt(/* @__PURE__ */ function() {
  return arguments;
}()) ? Qt : function(t) {
  return F(t) && Me.call(t, "callee") && !De.call(t, "callee");
};
function Ue() {
  return !1;
}
var Nr = typeof exports == "object" && exports && !exports.nodeType && exports, Zt = Nr && typeof module == "object" && module && !module.nodeType && module, Ve = Zt && Zt.exports === Nr, tr = Ve ? P.Buffer : void 0, Re = tr ? tr.isBuffer : void 0, q = Re || Ue, We = "[object Arguments]", Ge = "[object Array]", Je = "[object Boolean]", qe = "[object Date]", Ke = "[object Error]", Ye = "[object Function]", He = "[object Map]", Xe = "[object Number]", Qe = "[object Object]", Ze = "[object RegExp]", to = "[object Set]", ro = "[object String]", no = "[object WeakMap]", eo = "[object ArrayBuffer]", oo = "[object DataView]", io = "[object Float32Array]", ao = "[object Float64Array]", uo = "[object Int8Array]", co = "[object Int16Array]", fo = "[object Int32Array]", so = "[object Uint8Array]", lo = "[object Uint8ClampedArray]", po = "[object Uint16Array]", vo = "[object Uint32Array]", _ = {};
_[io] = _[ao] = _[uo] = _[co] = _[fo] = _[so] = _[lo] = _[po] = _[vo] = !0;
_[We] = _[Ge] = _[eo] = _[Je] = _[oo] = _[qe] = _[Ke] = _[Ye] = _[He] = _[Xe] = _[Qe] = _[Ze] = _[to] = _[ro] = _[no] = !1;
function bo(t) {
  return F(t) && Pt(t.length) && !!_[B(t)];
}
function zt(t) {
  return function(r) {
    return t(r);
  };
}
var $r = typeof exports == "object" && exports && !exports.nodeType && exports, G = $r && typeof module == "object" && module && !module.nodeType && module, yo = G && G.exports === $r, yt = yo && Er.process, U = function() {
  try {
    var t = G && G.require && G.require("util").types;
    return t || yt && yt.binding && yt.binding("util");
  } catch {
  }
}(), rr = U && U.isTypedArray, Ct = rr ? zt(rr) : bo, ho = Object.prototype, go = ho.hasOwnProperty;
function Lr(t, r) {
  var n = S(t), e = !n && ot(t), i = !n && !e && q(t), o = !n && !e && !i && Ct(t), a = n || e || i || o, u = a ? Le(t.length, String) : [], c = u.length;
  for (var f in t)
    (r || go.call(t, f)) && !(a && // Safari 9 has enumerable `arguments.length` in strict mode.
    (f == "length" || // Node.js 0.10 has enumerable non-index properties on buffers.
    i && (f == "offset" || f == "parent") || // PhantomJS 2 has enumerable non-index properties on typed arrays.
    o && (f == "buffer" || f == "byteLength" || f == "byteOffset") || // Skip index properties.
    Ft(f, c))) && u.push(f);
  return u;
}
function Br(t, r) {
  return function(n) {
    return t(r(n));
  };
}
var jo = Br(Object.keys, Object), mo = Object.prototype, _o = mo.hasOwnProperty;
function Oo(t) {
  if (!kt(t))
    return jo(t);
  var r = [];
  for (var n in Object(t))
    _o.call(t, n) && n != "constructor" && r.push(n);
  return r;
}
function Nt(t) {
  return ct(t) ? Lr(t) : Oo(t);
}
function wo(t) {
  var r = [];
  if (t != null)
    for (var n in Object(t))
      r.push(n);
  return r;
}
var Ao = Object.prototype, xo = Ao.hasOwnProperty;
function So(t) {
  if (!k(t))
    return wo(t);
  var r = kt(t), n = [];
  for (var e in t)
    e == "constructor" && (r || !xo.call(t, e)) || n.push(e);
  return n;
}
function $t(t) {
  return ct(t) ? Lr(t, !0) : So(t);
}
var To = zr(function(t, r, n, e) {
  kr(r, $t(r), t, e);
}), Eo = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, Io = /^\w*$/;
function Lt(t, r) {
  if (S(t))
    return !1;
  var n = typeof t;
  return n == "number" || n == "symbol" || n == "boolean" || t == null || Tt(t) ? !0 : Io.test(t) || !Eo.test(t) || r != null && t in Object(r);
}
var K = D(Object, "create");
function Fo() {
  this.__data__ = K ? K(null) : {}, this.size = 0;
}
function Po(t) {
  var r = this.has(t) && delete this.__data__[t];
  return this.size -= r ? 1 : 0, r;
}
var ko = "__lodash_hash_undefined__", zo = Object.prototype, Co = zo.hasOwnProperty;
function No(t) {
  var r = this.__data__;
  if (K) {
    var n = r[t];
    return n === ko ? void 0 : n;
  }
  return Co.call(r, t) ? r[t] : void 0;
}
var $o = Object.prototype, Lo = $o.hasOwnProperty;
function Bo(t) {
  var r = this.__data__;
  return K ? r[t] !== void 0 : Lo.call(r, t);
}
var Mo = "__lodash_hash_undefined__";
function Do(t, r) {
  var n = this.__data__;
  return this.size += this.has(t) ? 0 : 1, n[t] = K && r === void 0 ? Mo : r, this;
}
function L(t) {
  var r = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++r < n; ) {
    var e = t[r];
    this.set(e[0], e[1]);
  }
}
L.prototype.clear = Fo;
L.prototype.delete = Po;
L.prototype.get = No;
L.prototype.has = Bo;
L.prototype.set = Do;
function Uo() {
  this.__data__ = [], this.size = 0;
}
function ft(t, r) {
  for (var n = t.length; n--; )
    if (H(t[n][0], r))
      return n;
  return -1;
}
var Vo = Array.prototype, Ro = Vo.splice;
function Wo(t) {
  var r = this.__data__, n = ft(r, t);
  if (n < 0)
    return !1;
  var e = r.length - 1;
  return n == e ? r.pop() : Ro.call(r, n, 1), --this.size, !0;
}
function Go(t) {
  var r = this.__data__, n = ft(r, t);
  return n < 0 ? void 0 : r[n][1];
}
function Jo(t) {
  return ft(this.__data__, t) > -1;
}
function qo(t, r) {
  var n = this.__data__, e = ft(n, t);
  return e < 0 ? (++this.size, n.push([t, r])) : n[e][1] = r, this;
}
function z(t) {
  var r = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++r < n; ) {
    var e = t[r];
    this.set(e[0], e[1]);
  }
}
z.prototype.clear = Uo;
z.prototype.delete = Wo;
z.prototype.get = Go;
z.prototype.has = Jo;
z.prototype.set = qo;
var Y = D(P, "Map");
function Ko() {
  this.size = 0, this.__data__ = {
    hash: new L(),
    map: new (Y || z)(),
    string: new L()
  };
}
function Yo(t) {
  var r = typeof t;
  return r == "string" || r == "number" || r == "symbol" || r == "boolean" ? t !== "__proto__" : t === null;
}
function st(t, r) {
  var n = t.__data__;
  return Yo(r) ? n[typeof r == "string" ? "string" : "hash"] : n.map;
}
function Ho(t) {
  var r = st(this, t).delete(t);
  return this.size -= r ? 1 : 0, r;
}
function Xo(t) {
  return st(this, t).get(t);
}
function Qo(t) {
  return st(this, t).has(t);
}
function Zo(t, r) {
  var n = st(this, t), e = n.size;
  return n.set(t, r), this.size += n.size == e ? 0 : 1, this;
}
function C(t) {
  var r = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++r < n; ) {
    var e = t[r];
    this.set(e[0], e[1]);
  }
}
C.prototype.clear = Ko;
C.prototype.delete = Ho;
C.prototype.get = Xo;
C.prototype.has = Qo;
C.prototype.set = Zo;
var ti = "Expected a function";
function Bt(t, r) {
  if (typeof t != "function" || r != null && typeof r != "function")
    throw new TypeError(ti);
  var n = function() {
    var e = arguments, i = r ? r.apply(this, e) : e[0], o = n.cache;
    if (o.has(i))
      return o.get(i);
    var a = t.apply(this, e);
    return n.cache = o.set(i, a) || o, a;
  };
  return n.cache = new (Bt.Cache || C)(), n;
}
Bt.Cache = C;
var ri = 500;
function ni(t) {
  var r = Bt(t, function(e) {
    return n.size === ri && n.clear(), e;
  }), n = r.cache;
  return r;
}
var ei = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g, oi = /\\(\\)?/g, ii = ni(function(t) {
  var r = [];
  return t.charCodeAt(0) === 46 && r.push(""), t.replace(ei, function(n, e, i, o) {
    r.push(i ? o.replace(oi, "$1") : e || n);
  }), r;
});
function ai(t) {
  return t == null ? "" : Fr(t);
}
function Mr(t, r) {
  return S(t) ? t : Lt(t, r) ? [t] : ii(ai(t));
}
function lt(t) {
  if (typeof t == "string" || Tt(t))
    return t;
  var r = t + "";
  return r == "0" && 1 / t == -1 / 0 ? "-0" : r;
}
function Dr(t, r) {
  r = Mr(r, t);
  for (var n = 0, e = r.length; t != null && n < e; )
    t = t[lt(r[n++])];
  return n && n == e ? t : void 0;
}
function ui(t, r, n) {
  var e = t == null ? void 0 : Dr(t, r);
  return e === void 0 ? n : e;
}
function ci(t, r) {
  for (var n = -1, e = r.length, i = t.length; ++n < e; )
    t[i + n] = r[n];
  return t;
}
var Ur = Br(Object.getPrototypeOf, Object), fi = "[object Object]", si = Function.prototype, li = Object.prototype, Vr = si.toString, pi = li.hasOwnProperty, di = Vr.call(Object);
function vi(t) {
  if (!F(t) || B(t) != fi)
    return !1;
  var r = Ur(t);
  if (r === null)
    return !0;
  var n = pi.call(r, "constructor") && r.constructor;
  return typeof n == "function" && n instanceof n && Vr.call(n) == di;
}
function bi() {
  this.__data__ = new z(), this.size = 0;
}
function yi(t) {
  var r = this.__data__, n = r.delete(t);
  return this.size = r.size, n;
}
function hi(t) {
  return this.__data__.get(t);
}
function gi(t) {
  return this.__data__.has(t);
}
var ji = 200;
function mi(t, r) {
  var n = this.__data__;
  if (n instanceof z) {
    var e = n.__data__;
    if (!Y || e.length < ji - 1)
      return e.push([t, r]), this.size = ++n.size, this;
    n = this.__data__ = new C(e);
  }
  return n.set(t, r), this.size = n.size, this;
}
function E(t) {
  var r = this.__data__ = new z(t);
  this.size = r.size;
}
E.prototype.clear = bi;
E.prototype.delete = yi;
E.prototype.get = hi;
E.prototype.has = gi;
E.prototype.set = mi;
var Rr = typeof exports == "object" && exports && !exports.nodeType && exports, nr = Rr && typeof module == "object" && module && !module.nodeType && module, _i = nr && nr.exports === Rr, er = _i ? P.Buffer : void 0, or = er ? er.allocUnsafe : void 0;
function Wr(t, r) {
  if (r)
    return t.slice();
  var n = t.length, e = or ? or(n) : new t.constructor(n);
  return t.copy(e), e;
}
function Oi(t, r) {
  for (var n = -1, e = t == null ? 0 : t.length, i = 0, o = []; ++n < e; ) {
    var a = t[n];
    r(a, n, t) && (o[i++] = a);
  }
  return o;
}
function wi() {
  return [];
}
var Ai = Object.prototype, xi = Ai.propertyIsEnumerable, ir = Object.getOwnPropertySymbols, Si = ir ? function(t) {
  return t == null ? [] : (t = Object(t), Oi(ir(t), function(r) {
    return xi.call(t, r);
  }));
} : wi;
function Ti(t, r, n) {
  var e = r(t);
  return S(t) ? e : ci(e, n(t));
}
function mt(t) {
  return Ti(t, Nt, Si);
}
var _t = D(P, "DataView"), Ot = D(P, "Promise"), wt = D(P, "Set"), ar = "[object Map]", Ei = "[object Object]", ur = "[object Promise]", cr = "[object Set]", fr = "[object WeakMap]", sr = "[object DataView]", Ii = M(_t), Fi = M(Y), Pi = M(Ot), ki = M(wt), zi = M(jt), T = B;
(_t && T(new _t(new ArrayBuffer(1))) != sr || Y && T(new Y()) != ar || Ot && T(Ot.resolve()) != ur || wt && T(new wt()) != cr || jt && T(new jt()) != fr) && (T = function(t) {
  var r = B(t), n = r == Ei ? t.constructor : void 0, e = n ? M(n) : "";
  if (e)
    switch (e) {
      case Ii:
        return sr;
      case Fi:
        return ar;
      case Pi:
        return ur;
      case ki:
        return cr;
      case zi:
        return fr;
    }
  return r;
});
var Ci = Object.prototype, Ni = Ci.hasOwnProperty;
function $i(t) {
  var r = t.length, n = new t.constructor(r);
  return r && typeof t[0] == "string" && Ni.call(t, "index") && (n.index = t.index, n.input = t.input), n;
}
var it = P.Uint8Array;
function Mt(t) {
  var r = new t.constructor(t.byteLength);
  return new it(r).set(new it(t)), r;
}
function Li(t, r) {
  var n = Mt(t.buffer);
  return new t.constructor(n, t.byteOffset, t.byteLength);
}
var Bi = /\w*$/;
function Mi(t) {
  var r = new t.constructor(t.source, Bi.exec(t));
  return r.lastIndex = t.lastIndex, r;
}
var lr = I ? I.prototype : void 0, pr = lr ? lr.valueOf : void 0;
function Di(t) {
  return pr ? Object(pr.call(t)) : {};
}
function Gr(t, r) {
  var n = r ? Mt(t.buffer) : t.buffer;
  return new t.constructor(n, t.byteOffset, t.length);
}
var Ui = "[object Boolean]", Vi = "[object Date]", Ri = "[object Map]", Wi = "[object Number]", Gi = "[object RegExp]", Ji = "[object Set]", qi = "[object String]", Ki = "[object Symbol]", Yi = "[object ArrayBuffer]", Hi = "[object DataView]", Xi = "[object Float32Array]", Qi = "[object Float64Array]", Zi = "[object Int8Array]", ta = "[object Int16Array]", ra = "[object Int32Array]", na = "[object Uint8Array]", ea = "[object Uint8ClampedArray]", oa = "[object Uint16Array]", ia = "[object Uint32Array]";
function aa(t, r, n) {
  var e = t.constructor;
  switch (r) {
    case Yi:
      return Mt(t);
    case Ui:
    case Vi:
      return new e(+t);
    case Hi:
      return Li(t);
    case Xi:
    case Qi:
    case Zi:
    case ta:
    case ra:
    case na:
    case ea:
    case oa:
    case ia:
      return Gr(t, n);
    case Ri:
      return new e();
    case Wi:
    case qi:
      return new e(t);
    case Gi:
      return Mi(t);
    case Ji:
      return new e();
    case Ki:
      return Di(t);
  }
}
function Jr(t) {
  return typeof t.constructor == "function" && !kt(t) ? he(Ur(t)) : {};
}
var ua = "[object Map]";
function ca(t) {
  return F(t) && T(t) == ua;
}
var dr = U && U.isMap, fa = dr ? zt(dr) : ca, sa = "[object Set]";
function la(t) {
  return F(t) && T(t) == sa;
}
var vr = U && U.isSet, pa = vr ? zt(vr) : la, da = 1, qr = "[object Arguments]", va = "[object Array]", ba = "[object Boolean]", ya = "[object Date]", ha = "[object Error]", Kr = "[object Function]", ga = "[object GeneratorFunction]", ja = "[object Map]", ma = "[object Number]", Yr = "[object Object]", _a = "[object RegExp]", Oa = "[object Set]", wa = "[object String]", Aa = "[object Symbol]", xa = "[object WeakMap]", Sa = "[object ArrayBuffer]", Ta = "[object DataView]", Ea = "[object Float32Array]", Ia = "[object Float64Array]", Fa = "[object Int8Array]", Pa = "[object Int16Array]", ka = "[object Int32Array]", za = "[object Uint8Array]", Ca = "[object Uint8ClampedArray]", Na = "[object Uint16Array]", $a = "[object Uint32Array]", m = {};
m[qr] = m[va] = m[Sa] = m[Ta] = m[ba] = m[ya] = m[Ea] = m[Ia] = m[Fa] = m[Pa] = m[ka] = m[ja] = m[ma] = m[Yr] = m[_a] = m[Oa] = m[wa] = m[Aa] = m[za] = m[Ca] = m[Na] = m[$a] = !0;
m[ha] = m[Kr] = m[xa] = !1;
function Z(t, r, n, e, i, o) {
  var a, u = r & da;
  if (a !== void 0)
    return a;
  if (!k(t))
    return t;
  var c = S(t);
  if (c)
    a = $i(t);
  else {
    var f = T(t), s = f == Kr || f == ga;
    if (q(t))
      return Wr(t, u);
    if (f == Yr || f == qr || s && !i)
      a = s ? {} : Jr(t);
    else {
      if (!m[f])
        return i ? t : {};
      a = aa(t, f, u);
    }
  }
  o || (o = new E());
  var d = o.get(t);
  if (d)
    return d;
  o.set(t, a), pa(t) ? t.forEach(function(p) {
    a.add(Z(p, r, n, p, t, o));
  }) : fa(t) && t.forEach(function(p, j) {
    a.set(j, Z(p, r, n, j, t, o));
  });
  var v = mt, h = c ? void 0 : v(t);
  return Te(h || t, function(p, j) {
    h && (j = p, p = t[j]), Pr(a, j, Z(p, r, n, j, t, o));
  }), a;
}
var La = 1, Ba = 4;
function Ma(t) {
  return Z(t, La | Ba);
}
var Da = "__lodash_hash_undefined__";
function Ua(t) {
  return this.__data__.set(t, Da), this;
}
function Va(t) {
  return this.__data__.has(t);
}
function at(t) {
  var r = -1, n = t == null ? 0 : t.length;
  for (this.__data__ = new C(); ++r < n; )
    this.add(t[r]);
}
at.prototype.add = at.prototype.push = Ua;
at.prototype.has = Va;
function Ra(t, r) {
  for (var n = -1, e = t == null ? 0 : t.length; ++n < e; )
    if (r(t[n], n, t))
      return !0;
  return !1;
}
function Wa(t, r) {
  return t.has(r);
}
var Ga = 1, Ja = 2;
function Hr(t, r, n, e, i, o) {
  var a = n & Ga, u = t.length, c = r.length;
  if (u != c && !(a && c > u))
    return !1;
  var f = o.get(t), s = o.get(r);
  if (f && s)
    return f == r && s == t;
  var d = -1, v = !0, h = n & Ja ? new at() : void 0;
  for (o.set(t, r), o.set(r, t); ++d < u; ) {
    var p = t[d], j = r[d];
    if (e)
      var A = a ? e(j, p, d, r, t, o) : e(p, j, d, t, r, o);
    if (A !== void 0) {
      if (A)
        continue;
      v = !1;
      break;
    }
    if (h) {
      if (!Ra(r, function(l, y) {
        if (!Wa(h, y) && (p === l || i(p, l, n, e, o)))
          return h.push(y);
      })) {
        v = !1;
        break;
      }
    } else if (!(p === j || i(p, j, n, e, o))) {
      v = !1;
      break;
    }
  }
  return o.delete(t), o.delete(r), v;
}
function qa(t) {
  var r = -1, n = Array(t.size);
  return t.forEach(function(e, i) {
    n[++r] = [i, e];
  }), n;
}
function Ka(t) {
  var r = -1, n = Array(t.size);
  return t.forEach(function(e) {
    n[++r] = e;
  }), n;
}
var Ya = 1, Ha = 2, Xa = "[object Boolean]", Qa = "[object Date]", Za = "[object Error]", tu = "[object Map]", ru = "[object Number]", nu = "[object RegExp]", eu = "[object Set]", ou = "[object String]", iu = "[object Symbol]", au = "[object ArrayBuffer]", uu = "[object DataView]", br = I ? I.prototype : void 0, ht = br ? br.valueOf : void 0;
function cu(t, r, n, e, i, o, a) {
  switch (n) {
    case uu:
      if (t.byteLength != r.byteLength || t.byteOffset != r.byteOffset)
        return !1;
      t = t.buffer, r = r.buffer;
    case au:
      return !(t.byteLength != r.byteLength || !o(new it(t), new it(r)));
    case Xa:
    case Qa:
    case ru:
      return H(+t, +r);
    case Za:
      return t.name == r.name && t.message == r.message;
    case nu:
    case ou:
      return t == r + "";
    case tu:
      var u = qa;
    case eu:
      var c = e & Ya;
      if (u || (u = Ka), t.size != r.size && !c)
        return !1;
      var f = a.get(t);
      if (f)
        return f == r;
      e |= Ha, a.set(t, r);
      var s = Hr(u(t), u(r), e, i, o, a);
      return a.delete(t), s;
    case iu:
      if (ht)
        return ht.call(t) == ht.call(r);
  }
  return !1;
}
var fu = 1, su = Object.prototype, lu = su.hasOwnProperty;
function pu(t, r, n, e, i, o) {
  var a = n & fu, u = mt(t), c = u.length, f = mt(r), s = f.length;
  if (c != s && !a)
    return !1;
  for (var d = c; d--; ) {
    var v = u[d];
    if (!(a ? v in r : lu.call(r, v)))
      return !1;
  }
  var h = o.get(t), p = o.get(r);
  if (h && p)
    return h == r && p == t;
  var j = !0;
  o.set(t, r), o.set(r, t);
  for (var A = a; ++d < c; ) {
    v = u[d];
    var l = t[v], y = r[v];
    if (e)
      var w = a ? e(y, l, v, r, t, o) : e(l, y, v, t, r, o);
    if (!(w === void 0 ? l === y || i(l, y, n, e, o) : w)) {
      j = !1;
      break;
    }
    A || (A = v == "constructor");
  }
  if (j && !A) {
    var b = t.constructor, O = r.constructor;
    b != O && "constructor" in t && "constructor" in r && !(typeof b == "function" && b instanceof b && typeof O == "function" && O instanceof O) && (j = !1);
  }
  return o.delete(t), o.delete(r), j;
}
var du = 1, yr = "[object Arguments]", hr = "[object Array]", X = "[object Object]", vu = Object.prototype, gr = vu.hasOwnProperty;
function bu(t, r, n, e, i, o) {
  var a = S(t), u = S(r), c = a ? hr : T(t), f = u ? hr : T(r);
  c = c == yr ? X : c, f = f == yr ? X : f;
  var s = c == X, d = f == X, v = c == f;
  if (v && q(t)) {
    if (!q(r))
      return !1;
    a = !0, s = !1;
  }
  if (v && !s)
    return o || (o = new E()), a || Ct(t) ? Hr(t, r, n, e, i, o) : cu(t, r, c, n, e, i, o);
  if (!(n & du)) {
    var h = s && gr.call(t, "__wrapped__"), p = d && gr.call(r, "__wrapped__");
    if (h || p) {
      var j = h ? t.value() : t, A = p ? r.value() : r;
      return o || (o = new E()), i(j, A, n, e, o);
    }
  }
  return v ? (o || (o = new E()), pu(t, r, n, e, i, o)) : !1;
}
function Dt(t, r, n, e, i) {
  return t === r ? !0 : t == null || r == null || !F(t) && !F(r) ? t !== t && r !== r : bu(t, r, n, e, Dt, i);
}
var yu = 1, hu = 2;
function gu(t, r, n, e) {
  var i = n.length, o = i;
  if (t == null)
    return !o;
  for (t = Object(t); i--; ) {
    var a = n[i];
    if (a[2] ? a[1] !== t[a[0]] : !(a[0] in t))
      return !1;
  }
  for (; ++i < o; ) {
    a = n[i];
    var u = a[0], c = t[u], f = a[1];
    if (a[2]) {
      if (c === void 0 && !(u in t))
        return !1;
    } else {
      var s = new E(), d;
      if (!(d === void 0 ? Dt(f, c, yu | hu, e, s) : d))
        return !1;
    }
  }
  return !0;
}
function Xr(t) {
  return t === t && !k(t);
}
function ju(t) {
  for (var r = Nt(t), n = r.length; n--; ) {
    var e = r[n], i = t[e];
    r[n] = [e, i, Xr(i)];
  }
  return r;
}
function Qr(t, r) {
  return function(n) {
    return n == null ? !1 : n[t] === r && (r !== void 0 || t in Object(n));
  };
}
function mu(t) {
  var r = ju(t);
  return r.length == 1 && r[0][2] ? Qr(r[0][0], r[0][1]) : function(n) {
    return n === t || gu(n, t, r);
  };
}
function _u(t, r) {
  return t != null && r in Object(t);
}
function Ou(t, r, n) {
  r = Mr(r, t);
  for (var e = -1, i = r.length, o = !1; ++e < i; ) {
    var a = lt(r[e]);
    if (!(o = t != null && n(t, a)))
      break;
    t = t[a];
  }
  return o || ++e != i ? o : (i = t == null ? 0 : t.length, !!i && Pt(i) && Ft(a, i) && (S(t) || ot(t)));
}
function wu(t, r) {
  return t != null && Ou(t, r, _u);
}
var Au = 1, xu = 2;
function Su(t, r) {
  return Lt(t) && Xr(r) ? Qr(lt(t), r) : function(n) {
    var e = ui(n, t);
    return e === void 0 && e === r ? wu(n, t) : Dt(r, e, Au | xu);
  };
}
function Tu(t) {
  return function(r) {
    return r == null ? void 0 : r[t];
  };
}
function Eu(t) {
  return function(r) {
    return Dr(r, t);
  };
}
function Iu(t) {
  return Lt(t) ? Tu(lt(t)) : Eu(t);
}
function Fu(t) {
  return typeof t == "function" ? t : t == null ? Et : typeof t == "object" ? S(t) ? Su(t[0], t[1]) : mu(t) : Iu(t);
}
function Pu(t) {
  return function(r, n, e) {
    for (var i = -1, o = Object(r), a = e(r), u = a.length; u--; ) {
      var c = a[++i];
      if (n(o[c], c, o) === !1)
        break;
    }
    return r;
  };
}
var Zr = Pu();
function ku(t, r) {
  return t && Zr(t, r, Nt);
}
function At(t, r, n) {
  (n !== void 0 && !H(t[r], n) || n === void 0 && !(r in t)) && ut(t, r, n);
}
function zu(t) {
  return F(t) && ct(t);
}
function xt(t, r) {
  if (!(r === "constructor" && typeof t[r] == "function") && r != "__proto__")
    return t[r];
}
function Cu(t) {
  return kr(t, $t(t));
}
function Nu(t, r, n, e, i, o, a) {
  var u = xt(t, n), c = xt(r, n), f = a.get(c);
  if (f) {
    At(t, n, f);
    return;
  }
  var s = o ? o(u, c, n + "", t, r, a) : void 0, d = s === void 0;
  if (d) {
    var v = S(c), h = !v && q(c), p = !v && !h && Ct(c);
    s = c, v || h || p ? S(u) ? s = u : zu(u) ? s = je(u) : h ? (d = !1, s = Wr(c, !0)) : p ? (d = !1, s = Gr(c, !0)) : s = [] : vi(c) || ot(c) ? (s = u, ot(u) ? s = Cu(u) : (!k(u) || It(u)) && (s = Jr(c))) : d = !1;
  }
  d && (a.set(c, s), i(s, c, e, o, a), a.delete(c)), At(t, n, s);
}
function tn(t, r, n, e, i) {
  t !== r && Zr(r, function(o, a) {
    if (i || (i = new E()), k(o))
      Nu(t, r, a, n, tn, e, i);
    else {
      var u = e ? e(xt(t, a), o, a + "", t, r, i) : void 0;
      u === void 0 && (u = o), At(t, a, u);
    }
  }, $t);
}
var $u = zr(function(t, r, n, e) {
  tn(t, r, n, e);
});
function Lu(t, r) {
  var n = {};
  return r = Fu(r), ku(t, function(e, i, o) {
    ut(n, r(e, i, o), e);
  }), n;
}
const Bu = new RegExp("([\\p{Ll}\\d])(\\p{Lu})", "gu"), Mu = new RegExp("(\\p{Lu})([\\p{Lu}][\\p{Ll}])", "gu"), Du = new RegExp("(\\d)\\p{Ll}|(\\p{L})\\d", "u"), Uu = /[^\p{L}\d]+/giu, jr = "$1\0$2", mr = "";
function rn(t) {
  let r = t.trim();
  r = r.replace(Bu, jr).replace(Mu, jr), r = r.replace(Uu, "\0");
  let n = 0, e = r.length;
  for (; r.charAt(n) === "\0"; )
    n++;
  if (n === e)
    return [];
  for (; r.charAt(e - 1) === "\0"; )
    e--;
  return r.slice(n, e).split(/\0/g);
}
function Vu(t) {
  const r = rn(t);
  for (let n = 0; n < r.length; n++) {
    const e = r[n], i = Du.exec(e);
    if (i) {
      const o = i.index + (i[1] ?? i[2]).length;
      r.splice(n, 1, e.slice(0, o), e.slice(o));
    }
  }
  return r;
}
function Ru(t, r) {
  const [n, e, i] = qu(t, r), o = Wu(void 0), a = Gu(void 0), u = Ju(o, a);
  return n + e.map((c, f) => f === 0 ? o(c) : u(c, f)).join("") + i;
}
function Wu(t) {
  return (r) => r.toLocaleLowerCase(t);
}
function Gu(t) {
  return (r) => r.toLocaleUpperCase(t);
}
function Ju(t, r) {
  return (n, e) => {
    const i = n[0];
    return (e > 0 && i >= "0" && i <= "9" ? "_" + i : r(i)) + t(n.slice(1));
  };
}
function qu(t, r = {}) {
  const n = r.split ?? (r.separateNumbers ? Vu : rn), e = r.prefixCharacters ?? mr, i = r.suffixCharacters ?? mr;
  let o = 0, a = t.length;
  for (; o < t.length; ) {
    const u = t.charAt(o);
    if (!e.includes(u))
      break;
    o++;
  }
  for (; a > o; ) {
    const u = a - 1, c = t.charAt(u);
    if (!i.includes(c))
      break;
    a = u;
  }
  return [
    t.slice(0, o),
    n(t.slice(o, a)),
    t.slice(a)
  ];
}
const J = (t) => /.+-.+/.test(t) ? Ru(t) : t, Ku = (t) => Object.prototype.toString.call(t).slice(8, -1) === "Object";
function Yu(t, r) {
  const n = /* @__PURE__ */ Object.create(null), e = t.split(",");
  for (let i = 0; i < e.length; i++)
    n[e[i]] = !0;
  return (i) => !!n[i];
}
const Hu = /* @__PURE__ */ Yu("String,Number,Boolean,Function,Symbol,BigInt");
function Xu(t) {
  const r = t == null ? void 0 : t.toString().match(/^\s*function (\w+)/);
  return r ? r[1] : t === null ? "null" : "";
}
function Qu(t, r) {
  let n;
  const e = Xu(r);
  if (Hu(e)) {
    const i = typeof t;
    n = i === e.toLowerCase(), !n && i === "object" && (n = t instanceof r);
  } else e === "Object" ? n = k(t) : e === "Array" ? n = Array.isArray(t) : e === "null" ? n = t === null : n = t instanceof r;
  return {
    valid: n,
    expectedType: e
  };
}
function Zu({
  prop: t,
  type: r,
  validator: n
}) {
  if (![void 0, null].includes(t) && r) {
    let e = !1;
    const i = Array.isArray(r) ? r : [r], o = [];
    for (let a = 0; a < i.length && !e; a++) {
      const { valid: u, expectedType: c } = Qu(t, i[a]);
      o.push(c || ""), e = u;
    }
    if (!e)
      throw new TypeError(
        `Invalid prop: type check failed, expecting [${o.join(
          ", "
        )}], receiving: ${t}`
      );
  }
  if (n && !n(t))
    throw new Error(`Invalid prop: validator check failed, receiving: ${t}`);
}
function tc(t, {
  mergeObject: r,
  mergeObjectCustomizer: n,
  mergeFunction: e
}) {
  const i = [];
  for (let a = t.length - 1; a >= 0; a--)
    i.push(t[a]);
  const o = n || (e ? (a, u) => typeof a == "function" && typeof u == "function" ? e == null ? void 0 : e(u, a) : void 0 : void 0);
  return r === "deep" ? $u(...i, o) : To(...i, o);
}
function rc(t, {
  mergeFunction: r
}) {
  return t.reduce(r, () => {
  });
}
function $(t, r = {}) {
  const {
    type: n,
    default: e,
    defaultIsDynamic: i = !1,
    required: o = !1,
    validator: a,
    camelizeObjectKeys: u = !1,
    mergeObjectApplyOnlyToDefault: c = !1,
    mergeFunctionApplyOnlyToDefault: f = !0
  } = r;
  let { mergeObject: s = "deep", mergeObjectCustomizer: d, mergeFunction: v = !1 } = r;
  const h = [];
  let p, j = !1, A = !1;
  const l = (y) => {
    if (y !== void 0) {
      Zu({ type: n, prop: y, validator: a });
      const w = Ku(y), b = typeof y == "function";
      return j = w, A = b, w ? (y = Ma(y), u ? Lu(y, (O, g) => J(g)) : y) : y;
    }
  };
  for (const y of t)
    h.push(l(y));
  if (!i)
    h.push(l(e));
  else if (typeof e != "function")
    throw new TypeError(
      `Invalid option: options.default should be Function when options.defaultIsDynamic enabled, receiving: ${e}`
    );
  j || (s = !1, A || (v = !1));
  for (let y = 0; y < h.length; y++) {
    const w = h[y];
    if (w !== void 0) {
      y === h.length - 1 ? p = w : s ? p = tc(
        c ? [w, e] : h,
        {
          mergeObject: s,
          mergeObjectCustomizer: d,
          mergeFunction: v
        }
      ) : v ? p = rc(
        f ? [w, e] : h,
        {
          mergeFunction: v
        }
      ) : p = w;
      break;
    }
  }
  if (o && [void 0, null].includes(p))
    throw new Error("Missing required prop");
  return i ? $(t, {
    ...r,
    default: e(p),
    defaultIsDynamic: !1
  }) : p;
}
function _r(t) {
  const r = Array.from(J(t));
  return r[0] = r[0].toUpperCase(), r.unshift("o", "n"), r.join("");
}
function nc(t, {
  props: r = [],
  camelizePropNames: n = !1
} = {}) {
  const e = {
    props: {},
    attrs: {},
    listeners: {},
    hooks: {},
    slots: {}
  };
  let i;
  if (Array.isArray(r))
    i = n ? r.map((o) => J(o)) : r;
  else if (n) {
    i = [];
    for (const o in r)
      i.push(J(o));
  } else
    i = Object.keys(r);
  for (const o in t)
    if (o.startsWith("@")) {
      const a = o.substring(1);
      if (St)
        if (a.startsWith("vue:"))
          e.hooks[_r(a.replace("vue:", "vnode-"))] = t[o];
        else {
          if (a.startsWith("vnode"))
            throw new Error("@vnode-* hooks are no longer supported. Use the vue: prefix instead. For example, @vnode-mounted should be changed to @vue:mounted. @vnode-* hooks support has been removed in Vue 3.4.");
          e.listeners[_r(a)] = t[o];
        }
      else
        a.startsWith("hook:") ? e.hooks[a] = t[o] : e.listeners[a] = t[o];
    } else if (o.startsWith("#")) {
      const a = o.substring(1);
      a && (e.slots[a] = t[o]);
    } else {
      const a = n ? J(o) : o;
      i.includes(a) ? e.props[a] = t[o] : e.attrs[o] = t[o];
    }
  return e;
}
const Or = "JsonEditorVue", tt = [
  "mainMenuBar",
  "navigationBar",
  "statusBar",
  "askToFormat",
  "readOnly",
  "escapeControlCharacters",
  "escapeUnicodeCharacters",
  "flattenColumns"
], N = {}, wr = {}, rt = St ? "modelValue" : "value", Ar = St ? "update:modelValue" : "input", xr = {
  [rt]: {},
  mode: {
    type: String
  },
  debounce: {
    type: Number
  },
  stringified: {
    type: Boolean,
    default: void 0
  },
  ...Object.fromEntries(
    tt.map((t) => [
      t,
      {
        type: Boolean,
        default: void 0
      }
    ])
  )
}, ic = on({
  name: Or,
  install(t, r) {
    const n = nc(r || {}, { props: xr });
    Object.assign(N, n.props), Object.assign(wr, n.attrs), t.component(Or, this);
  },
  props: xr,
  emits: {
    [Ar](t) {
      return !0;
    },
    "update:mode": function(t) {
      return !0;
    }
  },
  setup(t, { attrs: r, emit: n, expose: e }) {
    var A;
    const i = (A = an()) == null ? void 0 : A.proxy, o = pt(), a = pt(!1), u = pt();
    un(() => {
      var l;
      u.value = $([t.mode, N.mode], {
        type: String
      }), (l = o.value) == null || l.updateProps({
        mode: u.value || nn.tree
      });
    });
    const c = (l) => {
      n("update:mode", l);
    };
    N.mode !== void 0 && t.mode === void 0 && c(N.mode);
    const f = Vt(() => $([t.debounce, N.debounce, 300], {
      type: Number
    })), s = Vt(() => $([t.stringified, N.stringified, !0], {
      type: Boolean
    }));
    let d = Q;
    const v = (l) => {
      a.value = !0, !s.value && l.text && (o.value && !o.value.validate() && (l.json = d(l.text)), l.text = void 0), n(
        Ar,
        l.text === void 0 ? l.json : l.text
      );
    }, h = Rn(v, f.value), p = (l) => {
      u.value === "text" ? h(l) : v(l);
    }, j = (l, y) => (...w) => {
      l(...w), y(...w);
    };
    return e == null || e({ jsonEditor: o }), cn(() => {
      var l;
      (l = o.value) == null || l.destroy();
    }), fn(() => {
      const l = $([t[rt], N[rt]]), y = Object.fromEntries(
        Array.from(tt, (b) => [b, $([t[b], N[b]])]).filter(
          ([, b]) => b !== void 0
        )
      ), w = $(
        [
          y,
          r,
          wr
        ],
        {
          camelizeObjectKeys: !0,
          defaultIsDynamic: !0,
          default: (b) => {
            var O;
            return d = ((O = b.parser) == null ? void 0 : O.parse) || Q, {
              onChange: p,
              onChangeMode: c,
              mode: u.value,
              // Can not just pass one of parse and stringify
              parser: {
                // SafeDestr is used by default so that it will not affect the result of jsonEditor.value.validate()
                // When stringified is disabled, destr is used by default for better performance (destr is only called when JSON is valid)
                parse: hn,
                stringify: JSON.stringify
              },
              ...l !== void 0 && {
                content: {
                  [typeof l == "string" && u.value === "text" && s.value ? "text" : "json"]: l
                }
              }
            };
          },
          mergeFunction: j,
          mergeObject: "shallow",
          type: Object
        }
      );
      o.value = en({
        target: i == null ? void 0 : i.$refs.jsonEditorRef,
        props: w
      }), dt(
        () => t[rt],
        (b) => {
          if (a.value) {
            a.value = !1;
            return;
          }
          o.value && o.value.set(
            [void 0, ""].includes(b) ? {
              text: ""
            } : {
              [typeof b == "string" && u.value === "text" && s.value ? "text" : "json"]: b
            }
          );
        },
        {
          deep: !0
        }
      ), dt(
        () => Array.from(tt, (b) => t[b]),
        (b) => {
          var O;
          (O = o.value) == null || O.updateProps(
            Object.fromEntries(Array.from(b, (g, x) => [tt[x], g]).filter(([, g]) => g !== void 0))
          );
        }
      ), dt(
        () => r,
        (b) => {
          var g, x;
          const O = {};
          (b.onChange || b["on-change"]) && (O.onChange = p), (b.onChangeMode || b["on-change-mode"]) && (O.onChangeMode = c), d = ((g = b.parser) == null ? void 0 : g.parse) || Q, (x = o.value) == null || x.updateProps(
            Object.getOwnPropertyNames(O).length > 0 ? $([b, O], {
              camelizeObjectKeys: !0,
              mergeFunction: j,
              mergeObject: "shallow",
              type: Object
            }) : b
          );
        },
        {
          deep: !0
        }
      ), e || (e = (b) => {
        for (const O in b)
          i[O] = sn(b[O]);
      }, e({ jsonEditor: o }));
    }), () => ln("div", { ref: "jsonEditorRef" });
  }
});
export {
  ic as default
};
